### PR TITLE
Weight decay sweep WD=2e-3, 5e-3: target OOD vol_p overfitting

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The val/test gap on `test_primary/volume_pressure_rel_l2_pct` (~3.0× ratio: val≈3.57%, test≈11.35%) strongly implicates overfitting to the training geometry distribution. The 4 OOD test cases expose a generalization deficit rather than a capacity deficit.

Weight decay is the most direct regularization lever available without any code changes. The current SOTA uses `--weight-decay 5e-4`, which was set early in the optimization sweep and may be too weak for a 400-sample dataset with OOD test cases. We have never explicitly swept WD — it has been at 5e-4 since PR #39.

**Hypothesis:** Increasing weight decay to 2e-3 or 5e-3 will penalize large weight magnitudes, reduce the model's ability to memorize training-set geometry, and close a fraction of the val/test vol_p gap. Even a 0.5× improvement on the 7.78pp val/test vol_p spread (3.57% → 11.35%) would be significant.

**Prior WD evidence:** The original Lion paper used WD=5e-3 for image classification on large-scale datasets. Our 400-sample automotive CFD dataset has a much higher per-sample information density per training iteration, making it closer to a "small-data high-complexity" regime where aggressive WD helps.

## Instructions

Run a **2-arm 4-epoch screen** using the PR #823 SOTA architecture exactly (no other changes).

### Arm A: WD = 2e-3

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward --optimizer lion --lr 9e-5 --weight-decay 2e-3 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --wandb-group edward-wd-reg-sweep \
  --wandb-name edward/wd-2e-3-screen
```

### Arm B: WD = 5e-3

Same as Arm A but `--weight-decay 5e-3`:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward --optimizer lion --lr 9e-5 --weight-decay 5e-3 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --wandb-group edward-wd-reg-sweep \
  --wandb-name edward/wd-5e-3-screen
```

### Sequencing

Run Arm A first. If it passes EP2 gate (≤16%), run Arm B in parallel or back-to-back. If Arm A fails EP1 (>30%), cancel Arm B and report.

## 4-Epoch Kill Gates (standard protocol)

| Epoch | Gate | Action if breached |
|---|---|---|
| EP1 | val_abupt ≤ 30% | Kill Arm, do not run next Arm |
| EP2 | val_abupt ≤ 16% | Kill if not met |
| EP3 | val_abupt ≤ 8.0% | Kill if not met |
| EP4 | val_abupt < 6.9% | Report result |

**Win criterion:** EP4 val_abupt < 6.9% (SOTA is 6.4407%) **AND** val→test vol_p ratio < 2.7× (SOTA is 3.03×). We want to see the val/test vol_p gap narrow, not just overall val improvement.

**Additional metric to report:** For each completed arm, please report:
- EP4 val: volume_pressure_rel_l2_pct
- EP4 test (if available): volume_pressure_rel_l2_pct
- Ratio: test_vol_p / val_vol_p

## Baseline

**Single-model SOTA:** PR #823 (nezuko), W&B run `ghh0s4ne`
- val_abupt = **6.4407%**
- test_abupt = **7.6992%**
- val vol_pressure = 3.8557%
- test vol_pressure = 11.6704%
- val/test vol_p ratio = **3.03×** ← this is what we want to reduce
- WD used = **5e-4** (current)

**Ensemble SOTA:** PR #880, val=6.0289%, test=7.3693%, test_vol_p=11.3478%

**Reproduce baseline (PR #823):**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent nezuko --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --wandb-group nezuko-surf-vol-xattn \
  --wandb-name nezuko/surf-vol-xattn-13ep
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Acknowledged — Arm A (WD=0.002) is running. I can see 8 runs in W&B group `edward-wd-reg-sweep` at step ~1,822 (one cohort crashed at step ~408 and was restarted successfully).

**Gate schedule (DDP=8, 5,493 steps/epoch):**
- EP1 ≤30% (~3,671 steps away)
- EP2 ≤16%
- EP3 ≤8.0%
- EP4 <6.9% (merge bar)

**Win criterion:** EP4 val_abupt <6.9% AND val→test vol_p ratio <2.7× (current baseline: WD=5e-4, val_abupt=6.4407%, vol_p ratio=3.03×)

**Baseline to beat:** PR #823 (nezuko, `ghh0s4ne`): val_abupt=6.4407%, test_abupt=7.6992%

When Arm A results are in at EP4, start Arm B (WD=0.005) per the PR instructions. Post gate results at each epoch boundary. Good luck.

**Comment by morganmcg1** [comment]
STUDENT edward:
SENPAI-RESULT: {"terminal":true,"status":"failed","pending_arms":false,"wandb_run_ids":["2xznrfmc"],"primary_metric":{"name":"val_abupt_axis_rel_l2_pct","value":30.88}}

## Results — Arm A killed at EP1 (gate breach), Arm B cancelled

**Decision: Arm A killed mid-EP2; Arm B cancelled per `If Arm A fails EP1 (>30%), cancel Arm B and report`.**

### Arm A: WD = 2e-3, EP1 metrics (from W&B `2xznrfmc`)

| Metric | Arm A (WD=2e-3) | EP1 gate | Status |
|---|---|---|---|
| **val_abupt_axis_rel_l2_pct** | **30.88%** | ≤30% | **BREACH (+0.88pp)** |
| val_surface_pressure_rel_l2_pct | 23.99% | — | — |
| val_volume_pressure_rel_l2_pct | 17.46% | — | — |
| val_wall_shear_rel_l2_pct | 34.43% | — | — |
| train_epoch_loss | 0.3758 | — | — |

### EP1 comparison vs recent 4ep SOTA-arch screens (same arch, lower WD)

| Run | WD | EP1 val_abupt |
|---|---|---|
| rff32-4ep | 5e-4 | 25.93% |
| gradnorm-a05-l5-sota | 5e-4 | 26.19% |
| no-qknorm-4ep | 5e-4 | 26.35% |
| drop-path-010 (Arm B) | 5e-4 | 28.49% |
| drop-path-005 (Arm A) | 5e-4 | 29.24% |
| **wd-2e-3 (this run)** | **2e-3** | **30.88%** ← worst, gate breach |

WD=2e-3 is ≥1.6pp worse at EP1 than every recent SOTA-arch baseline and 5.0pp worse than the cleanest comparable (rff32). This is well outside run-to-run noise (±0.5–1pp).

### Command run (Arm A)

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent edward --optimizer lion --lr 9e-5 --weight-decay 2e-3 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 4 --epochs 4 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --wandb-group edward-wd-reg-sweep \
  --wandb-name edward/wd-2e-3-screen
```

- W&B group: `edward-wd-reg-sweep`
- W&B run IDs (rank 0–7): `2xznrfmc, k52wn9vo, oqzrkiyw, vcimkbtj, 3h01lwwq, 196ij5i7, zvhpklej, vep5e3xn`
- Peak GPU memory at val time: ~49.77 GiB reserved (rank 0)
- EP1 wallclock: 5,992 s (~100 min) for 8× H100
- Process killed at step ~461/10864 of EP2 (~4% into EP2) at 2026-05-09 ~21:20 UTC

### What happened

The hypothesis predicted that higher WD would slow early training but improve generalization at later epochs. The first half of that prediction is now confirmed: WD=2e-3 produces the **worst EP1 val_abupt of any recent SOTA-arch 4ep screen**, by a margin (+1.6pp to +5.0pp) that is well outside run-to-run noise.

Whether the trade-off "buys" generalization improvement at EP3–EP4 is **untestable from this kill-gate protocol** — we never reached EP3 or EP4. The val/test vol_p ratio (the actual hypothesis target) requires test eval at EP4, which never happened.

Per the standard kill protocol (`EP1 | val_abupt ≤ 30% | Kill Arm, do not run next Arm`) and PR-specific sequencing (`If Arm A fails EP1 (>30%), cancel Arm B and report`), Arm A was terminated and Arm B (WD=5e-3) was not launched. WD=5e-3 would almost certainly fail EP1 even harder given its 2.5× stronger regularization.

### Suggested follow-ups

1. **Test the hypothesis with a less aggressive WD step.** Try `WD=1e-3` (2× current, half of Arm A) — same architecture, 4ep screen. This is the smallest meaningful step up from baseline 5e-4 and is more likely to clear EP1.
2. **Schedule WD instead of constant.** Use a low WD (5e-4) for the first 1–2 epochs while the model fits the bulk of the variance, then ramp to 1e-3 or 2e-3 for the late epochs where overfitting on OOD vol_p actually appears. This separates "learning capacity" (early) from "generalization regularization" (late) and aligns with the empirical fact that the val/test vol_p gap widens at late epochs.
3. **WD on selected parameter groups only.** Apply higher WD (2e-3) only to the volume decoder / vol-x-attn branch where vol_p memorization presumably occurs, while keeping surface decoder + backbone at 5e-4. This targets the diagnosed overfitting more surgically.
4. **Pair with longer training (13ep) at WD=1e-3.** A 4ep screen is a poor harness for a regularization hypothesis whose payoff is at the *end* of training. If WD=1e-3 passes EP1–EP2 in a 4ep screen, run a full 13ep confirmation arm to actually measure val/test vol_p ratio against the 3.03× SOTA baseline.
5. **Check whether Lion needs less WD than AdamW.** Lion's update magnitude is bounded by sign(...), so the *effective* WD relative to gradient size is already higher than for AdamW at the same nominal value. The Lion paper's WD=5e-3 was on 1B-image-scale data; for our 400-sample regime the equivalent may be much lower. This is consistent with EP1 catastrophically slowing down at WD=2e-3.

### Key take-away

WD=2e-3 over-regularizes a 400-sample DrivAerML run badly enough that early-training learning is impaired beyond the kill gate. The hypothesis (WD reduces val/test vol_p gap) remains untested. Recommend smaller WD steps (1e-3) and/or scheduled WD for any follow-up sweep.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #922 because WD=2e-3 over-regularizes a 400-sample Lion run — EP1 val_abupt=30.88% (gate is ≤30%), 5.0pp worse than the cleanest baseline and well outside noise. WD=5e-3 cancelled. The WD hypothesis remains untested at this scale; follow-up should start at WD=1e-3 (2× current). Closing as a dead end. Excellent analysis in the results — the student's suggested follow-ups are sound. edward is now idle and will be assigned a new experiment..
